File: PortProject.Api/Domain/DataRightsAggregate/DataRightsRequest.cs
namespace PortProject.Api.Domain.DataRightsAggregate;

public class DataRightsRequest
{
    public Guid Id { get; private set; }
    public string UserEmail { get; private set; }
    public DataRightsRequestType RequestType { get; private set; }
    public string? Details { get; private set; }
    public DataRightsRequestStatus Status { get; private set; }
    public DateTime RequestedAt { get; private set; }
    public DateTime? ProcessedAt { get; private set; }
    public string? ProcessedBy { get; private set; }
    public string? Response { get; private set; }

    private DataRightsRequest() { } // For EF Core

    public DataRightsRequest(string userEmail, DataRightsRequestType requestType, string? details = null)
    {
        if (string.IsNullOrWhiteSpace(userEmail))
            throw new ArgumentException("User email is required.", nameof(userEmail));

        Id = Guid.NewGuid();
        UserEmail = userEmail.ToLowerInvariant();
        RequestType = requestType;
        Details = details;
        Status = DataRightsRequestStatus.Pending;
        RequestedAt = DateTime.UtcNow;
    }

    public void MarkAsProcessed(string processedBy, string? response = null)
    {
        Status = DataRightsRequestStatus.Processed;
        ProcessedAt = DateTime.UtcNow;
        ProcessedBy = processedBy;
        Response = response;
    }

    public void MarkAsRejected(string processedBy, string reason)
    {
        Status = DataRightsRequestStatus.Rejected;
        ProcessedAt = DateTime.UtcNow;
        ProcessedBy = processedBy;
        Response = reason;
    }
}

public enum DataRightsRequestType
{
    DataAccess,      // Request a copy of personal data
    DataRectification, // Request correction of data
    DataDeletion     // Request deletion of data
}

public enum DataRightsRequestStatus
{
    Pending,
    Processed,
    Rejected
}

File: port-spa-app/src/domain/dataRights/DataRightsRequest.ts
export const DataRequestType = {
    Access: 'DataAccess',
    Rectification: 'DataRectification',
    Erasure: 'DataDeletion'
} as const;

export type DataRequestType = typeof DataRequestType[keyof typeof DataRequestType];

export const DataRequestStatus = {
    Pending: 'Pending',
    Processed: 'Processed',
    Rejected: 'Rejected'
} as const;

export type DataRequestStatus = typeof DataRequestStatus[keyof typeof DataRequestStatus];

export interface DataRightsRequest {
    id: string;
    userEmail: string;
    requestType: DataRequestType;
    status: DataRequestStatus;
    details?: string;
    response?: string;
    requestedAt: string;
    processedAt?: string;
    processedBy?: string;
}

export interface CreateDataRightsRequest {
    requestType: DataRequestType;
    details?: string;
}

export interface UpdateDataRightsRequest {
    status: DataRequestStatus;
    response?: string;
}
